// testcafe/lib/test-run/execute-js-expression.js
'use strict';

exports.__esModule = true;
exports.executeJsExpression = executeJsExpression;

var _vm = require('vm');

var _selectorBuilder = require('../client-functions/selectors/selector-builder');

var _selectorBuilder2 = _interopRequireDefault(_selectorBuilder);

var _clientFunctionBuilder = require('../client-functions/client-function-builder');

var _clientFunctionBuilder2 = _interopRequireDefault(_clientFunctionBuilder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const contextsInfo = [];

function getContextInfo(testRun) {
    let contextInfo = contextsInfo.find(info => info.testRun === testRun);

    if (!contextInfo) {
        contextInfo = { testRun, context: createExecutionContext(testRun), options: {} };

        contextsInfo.push(contextInfo);
    }

    return contextInfo;
}

function getContext(testRun, options = {}) {
    const contextInfo = getContextInfo(testRun);

    contextInfo.options = options;

    return contextInfo.context;
}

function createExecutionContext(testRun) {
    const sandbox = {
        Selector: (fn, options = {}) => {
            var _getContextInfo$optio = getContextInfo(testRun).options;
            const skipVisibilityCheck = _getContextInfo$optio.skipVisibilityCheck,
                  collectionMode = _getContextInfo$optio.collectionMode;


            if (skipVisibilityCheck) options.visibilityCheck = false;

            if (testRun && testRun.id) options.boundTestRun = testRun;

            if (collectionMode) options.collectionMode = collectionMode;

            const builder = new _selectorBuilder2.default(fn, options, { instantiation: 'Selector' });

            return builder.getFunction();
        },

        ClientFunction: (fn, options = {}) => {
            if (testRun && testRun.id) options.boundTestRun = testRun;

            const builder = new _clientFunctionBuilder2.default(fn, options, { instantiation: 'ClientFunction' });

            return builder.getFunction();
        }
    };

    return (0, _vm.createContext)(sandbox);
}

function executeJsExpression(expression, testRun, options) {
    const context = getContext(testRun, options);

    return (0, _vm.runInContext)(expression, context, { displayErrors: false });
}
